src/azul/plugins/__init__.py

```diff
@@ -206,33 +207,33 @@ def list_sources(self,
         """
         raise NotImplementedError
 
-    def resolve_source(self,
-                       *,
-                       spec: str,
-                       id: Optional[str] = None,
-                       verify_id: bool = True) -> SOURCE_REF:
+    @cached_property
+    def _source_ref_cls(self) -> Type[SOURCE_REF]:
+        cls = type(self)
+        base_cls = one(getattr(cls, '__orig_bases__'))
+        spec_cls, ref_cls = get_args(base_cls)
+        require(issubclass(spec_cls, SourceSpec))
+        require(issubclass(ref_cls, SourceRef))
+        assert ref_cls.spec_cls() is spec_cls
+        return ref_cls
+
+    def source_from_json(self, ref: JSON) -> SOURCE_REF:
+        return self._source_ref_cls.from_json(ref)
+
+    def resolve_source(self, spec: str) -> SOURCE_REF:
         """
         Return an instance of :class:`SourceRef` for the repository source
         matching the given specification or raise an exception if no such source
-        exists. If an ID is given and `verify_id` is set to True, ensure that
-        the source matching the specification has the given ID.
+        exists.
         """
-        cls = type(self)
-        base_cls = one(getattr(cls, '__orig_bases__'))
-        source_spec_cls, source_ref_cls = get_args(base_cls)
-        require(issubclass(source_spec_cls, SourceSpec))
-        require(issubclass(source_ref_cls, SourceRef))
-        spec = source_spec_cls.parse(spec)
-        if verify_id:
-            actual_id = self.lookup_source_id(spec)
-            if id is None:
-                id = actual_id
-            else:
-                require(id == actual_id,
-                        'Source ID changed unexpectedly', spec, id, actual_id)
-        else:
-            require(id is not None)
-        return source_ref_cls(id=id, spec=spec)
+        ref_cls = self._source_ref_cls
+        spec = ref_cls.spec_cls().parse(spec)
+        id = self.lookup_source_id(spec)
+        return ref_cls(id=id, spec=spec)
```

Comment:
Do you also get a PyCharm warning here?

Reply:
I do not.

Reply:
![2021-06-24-17:17:22](https://user-images.githubusercontent.com/48401926/123350581-24e34800-d510-11eb-8e8d-25a9f6bf0675.png)

Reply:
I get that, too. PyCharm doesn't realize that `@cached_property` should be treated like `@property`.